v1/src/test/java/com/google/cloud/teleport/templates/PubSubSubscriptionToBigQueryIT.java

```diff
@@ -0,0 +1,139 @@
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Integration test for {@link PubSubToBigQuery} PubSub Subscription to Bigquery. */
+@Category(TemplateIntegrationTest.class)
+@TemplateIntegrationTest(
+    value = PubSubToBigQuery.class,
+    template = "PubSub_Subscription_to_BigQuery")
+@RunWith(JUnit4.class)
+public class PubSubSubscriptionToBigQueryIT extends TemplateTestBase {
+  private static PubsubResourceManager pubsubResourceManager;
+  private static BigQueryResourceManager bigQueryResourceManager;
+
+  @Before
+  public void setUp() throws IOException {
+    pubsubResourceManager =
+        DefaultPubsubResourceManager.builder(testName.getMethodName(), PROJECT)
+            .credentialsProvider(credentialsProvider)
+            .build();
+    bigQueryResourceManager =
+        DefaultBigQueryResourceManager.builder(testName.getMethodName(), PROJECT)
+            .setCredentials(credentials)
+            .build();
+  }
+
+  @After
+  public void cleanUp() {
+    pubsubResourceManager.cleanupAll();
+    bigQueryResourceManager.cleanupAll();
+  }
+
+  @Test
+  public void testSubscriptionToBigQuery() throws IOException {
+    // Arrange
+    String jobName = createJobName(testName.getMethodName());
+    String bqTable = testName.getMethodName();
+    List<Map<String, Object>> messages =
+        List.of(
+            Map.of("job", jobName, "msgNumber", 1),
+            Map.of("job", jobName, "msgNumber", 2),
+            Map.of("job", jobName, "msgNumber", 3));
+    List<Field> bqSchemaFields =
+        Arrays.asList(
+            Field.of("job", StandardSQLTypeName.STRING),
+            Field.of("msgNumber", StandardSQLTypeName.INT64));
+    Schema bqSchema = Schema.of(bqSchemaFields);
+
+    TopicName topic = pubsubResourceManager.createTopic("input-topic");
+    SubscriptionName subscription =
+        pubsubResourceManager.createSubscription(topic, "input-subscription");
+    publishMessages(topic, messages);
+
+    bigQueryResourceManager.createDataset(REGION);
+    bigQueryResourceManager.createTable(bqTable, bqSchema);
+    String tableSpec = PROJECT + ":" + bigQueryResourceManager.getDatasetId() + "." + bqTable;
+
+    LaunchConfig.Builder options =
+        LaunchConfig.builder(jobName, specPath)
+            .addParameter("inputSubscription", subscription.toString())
+            .addParameter("outputTableSpec", tableSpec);
+
+    // Act
+    JobInfo info = launchTemplate(options);
+    assertThat(info.state()).isIn(JobState.ACTIVE_STATES);
+
+    AtomicReference<TableResult> records = new AtomicReference<>();
+    Result result =
+        new DataflowOperator(getDataflowClient())
+            .waitForConditionAndFinish(
+                createConfig(info),
+                () -> {
+                  TableResult values = bigQueryResourceManager.readTable(bqTable);
+                  records.set(values);
+                  return values.getTotalRows() == messages.size();
```

Comment:
I'd take care about using `==` when using Pub/Sub, as it doesn't offer exactly once guarantees. Maybe we can `>=` (and in the future we may want to deduplicate in the assertions, but at least we'll get a faster failure and not wait for a condition that already passed, it it makes sense).

Reply:
Makes sense! I did not know Pub/Sub does not guarantee exactly once